src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala

```diff
@@ -142,4 +143,86 @@ class DeltaLakeRelation(spark: SparkSession, override val plan: LogicalRelation)
       /* modificationTime */ modificationTime,
       path)
   }
+
+  private def deltaLakeVersionHistory(index: IndexLogEntry): Seq[(Int, Int)] = {
+    val versions =
+      index.derivedDataset.properties.properties
+        .getOrElse(DeltaLakeConstants.DELTA_VERSION_HISTORY_PROPERTY, "")
+
+    versions.split(",").reverseIterator.foldLeft(Seq[(Int, Int)]()) { (list, versionStr) =>
+      val pair = versionStr.split(":")
+      val Array(indexLogVersion, deltaTableVersion) = versionStr.split(":").map(_.toInt)
+      if (list.nonEmpty && list.head._2 == deltaTableVersion) {
+        // Omit if the delta lake version is the same. In this case, we only take the higher
+        // index log version as this is usually from index optimizations.
+        list
+      } else {
+        (indexLogVersion, deltaTableVersion) +: list
+      }
+    }
+  }
+
+  /**
+   * Returns IndexLogEntry of the closest index version for the given relation.
+   *
+   * curFiles is used to calculate the similarity with each index version data.
+   * Delta Lake source provider utilizes the history information of DELTA_VERSION_HISTORY_PROPERTY
+   * to find the closet version of the index.
+   *
+   * @param curFiles List of FileInfo for the source files in the relation.
+   * @param index Candidate index to be applied.
+   * @return IndexLogEntry of the closest version among available index versions.
+   */
+  override def closestIndexVersion(
+      curFiles: Seq[FileInfo],
+      index: IndexLogEntry): IndexLogEntry = {
+    // Seq of ( index log version, delta lake table version )
+    val versions = deltaLakeVersionHistory(index)
+    lazy val indexManager = Hyperspace.getContext(spark).indexCollectionManager
+    def getIndexLogEntry(logVersion: Int): IndexLogEntry = {
+      indexManager
+        .getLogManager(index.name)
+        .get
+        .getLog(logVersion)
+        .get
+        .asInstanceOf[IndexLogEntry]
+    }
+
+    plan.relation match {
+      case HadoopFsRelation(location: TahoeLogFileIndex, _, _, _, _, _) =>
+        val equalOrLessLastIndex = versions.lastIndexWhere(location.tableVersion >= _._2)
+        if (equalOrLessLastIndex == versions.size - 1) {
+          // Use the latest version.
+          index
+        } else if (equalOrLessLastIndex == -1) {
+          getIndexLogEntry(versions.head._1)
+        } else if (versions(equalOrLessLastIndex)._2 == location.tableVersion) {
+          getIndexLogEntry(versions(equalOrLessLastIndex)._1)
+        } else {
```

Comment:
Could you add some comments for each block so that one can understand the logic without reading the code? 😄